### ISSUE TITLE
BLIP-2 OPT VQA Fine-Tuning Decodes Bad Text

### ISSUE THREAD

**Issue by aburns4** [salesforce/LAVIS#376]
Hi,

Now that BLIP-2 OPT supports transformers 4.27, I've updated this library and am trying to fine tune on a custom vqa dataset with BLIP-2 OPT. However, `predict_answers` is generating nonsensical outputs like "????". Is this related to the package version which blip2 previously did not support, or potentially the tokenizer padding? The `forward` and `prediction_answers` functions use different padding sides, which I am not sure is intentional/part of the problem. Changing this around did not fix the issue, though. 

The same task and dataset train normally with the FlanT5 model variant. Any suggestions on what could be causing this would be appreciated!

Thank you very much.

**Comment by TalhaUsuf** [comment]
> Hi,
> 
> Now that BLIP-2 OPT supports transformers 4.27, I've updated this library and am trying to fine tune on a custom vqa dataset with BLIP-2 OPT. However, `predict_answers` is generating nonsensical outputs like "????". Is this related to the package version which blip2 previously did not support, or potentially the tokenizer padding? The `forward` and `prediction_answers` functions use different padding sides, which I am not sure is intentional/part of the problem. Changing this around did not fix the issue, though. 
> 
> The same task and dataset train normally with the FlanT5 model variant. Any suggestions on what could be causing this would be appreciated!
> 
> Thank you very much.

Try changing the parameter ```temperature```

**Comment by aburns4** [comment]
Are you suggesting changing the temperature in the generate call [here](https://github.com/salesforce/LAVIS/blob/main/lavis/models/blip2_models/blip2_opt.py#L337)? I've tried setting higher and lower temperatures and it did not help- the val accuracy is still zero due to text outputs like "___" and "???"

**Comment by dszpr** [comment]
Hi！ @aburns4 
I noticed that in forward function in blip2_opt.py, only the questions in the VQA dateset are used. Both the text_input and the target lables are derived from the opt_tokens
`
        text = [t + "\n" for t in samples["text_input"]]

        opt_tokens = self.opt_tokenizer(
            text,
            return_tensors="pt",
            padding="longest",
            truncation=True,
            max_length=self.max_txt_len,
        ).to(image.device)

        targets = opt_tokens.input_ids.masked_fill(
            opt_tokens.input_ids == self.opt_tokenizer.pad_token_id, -100
        )
        if self.prompt:
            targets[:, : self.prompt_length] = -100  # do not apply loss to the prompt

        empty_targets = (
            torch.ones(atts_opt.size(), dtype=torch.long).to(image.device).fill_(-100)
        )
        targets = torch.cat([empty_targets, targets], dim=1)

        inputs_embeds = self.opt_model.model.decoder.embed_tokens(opt_tokens.input_ids)
        inputs_embeds = torch.cat([inputs_opt, inputs_embeds], dim=1)
        attention_mask = torch.cat([atts_opt, opt_tokens.attention_mask], dim=1)

        with self.maybe_autocast():
            outputs = self.opt_model(
                inputs_embeds=inputs_embeds,
                attention_mask=attention_mask,
                return_dict=True,
                labels=targets,
            )
`
In the VQA task,  are the target lables supposed to be the answers in the VQA dataset? But the answers in the VQA datasetare not used in the blip2_opt.py. However, the answers are used as target lables in blip2_t5.py. It really confused.
Have you make any change to blip2_opt.py?
